### DISCUSSION TITLE
Add custom NER pipeline to existing model

### DISCUSSION THREAD

**Discussion by adicrescenzo** [explosion/spaCy#8192] [Help: Coding & Implementations] [answered]
I'm using the model "it_core_news_lg" to perform the following tasks: part of speech tagging (POS), name entity recognition (NER) and syntactic dependency (DEP). 

I've a small dataset and I trained a NER pipeline to recognize some custom entities.
I would like to use the existing model "it_core_news_lg" to perform POS and DEP but I'm trying to replace the default NER pipeline with the custom one I've created.

This is the config file I used to train the NER pipeline:

    [paths]
    train = null
    dev = null
    vectors = null
    init_tok2vec = null
    
    [system]
    gpu_allocator = null
    seed = 0
    
    [nlp]
    lang = "it"
    pipeline = ["tok2vec","ner"]
    batch_size = 1000
    disabled = []
    before_creation = null
    after_creation = null
    after_pipeline_creation = null
    tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
    
    [components]
    
    [components.ner]
    factory = "ner"
    moves = null
    update_with_oracle_cut_size = 100
    
    [components.ner.model]
    @architectures = "spacy.TransitionBasedParser.v2"
    state_type = "ner"
    extra_state_tokens = false
    hidden_width = 64
    maxout_pieces = 2
    use_upper = true
    nO = null
    
    [components.ner.model.tok2vec]
    @architectures = "spacy.Tok2VecListener.v1"
    width = ${components.tok2vec.model.encode.width}
    upstream = "*"
    
    [components.tok2vec]
    factory = "tok2vec"
    
    [components.tok2vec.model]
    @architectures = "spacy.Tok2Vec.v2"
    
    [components.tok2vec.model.embed]
    @architectures = "spacy.MultiHashEmbed.v2"
    width = ${components.tok2vec.model.encode.width}
    attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
    rows = [5000,2500,2500,2500]
    include_static_vectors = true
    
    [components.tok2vec.model.encode]
    @architectures = "spacy.MaxoutWindowEncoder.v2"
    width = 256
    depth = 8
    window_size = 1
    maxout_pieces = 3
    
    [corpora]
    
    [corpora.dev]
    @readers = "spacy.Corpus.v1"
    path = ${paths.dev}
    max_length = 0
    gold_preproc = false
    limit = 0
    augmenter = null
    
    [corpora.train]
    @readers = "spacy.Corpus.v1"
    path = ${paths.train}
    max_length = 2000
    gold_preproc = false
    limit = 0
    augmenter = null
    
    [training]
    dev_corpus = "corpora.dev"
    train_corpus = "corpora.train"
    seed = ${system.seed}
    gpu_allocator = ${system.gpu_allocator}
    dropout = 0.1
    accumulate_gradient = 1
    patience = 1600
    max_epochs = 0
    max_steps = 20000
    eval_frequency = 200
    frozen_components = []
    before_to_disk = null
    
    [training.batcher]
    @batchers = "spacy.batch_by_words.v1"
    discard_oversize = false
    tolerance = 0.2
    get_length = null
    
    [training.batcher.size]
    @schedules = "compounding.v1"
    start = 100
    stop = 1000
    compound = 1.001
    t = 0.0
    
    [training.logger]
    @loggers = "spacy.ConsoleLogger.v1"
    progress_bar = false
    
    [training.optimizer]
    @optimizers = "Adam.v1"
    beta1 = 0.9
    beta2 = 0.999
    L2_is_weight_decay = true
    L2 = 0.01
    grad_clip = 1.0
    use_averages = false
    eps = 0.00000001
    learn_rate = 0.001
    
    [training.score_weights]
    ents_per_type = null
    ents_f = 1.0
    ents_p = 0.0
    ents_r = 0.0
    
    [pretraining]
    
    [initialize]
    vectors = "it_core_news_lg"
    init_tok2vec = ${paths.init_tok2vec}
    vocab_data = null
    lookups = null
    before_init = null
    after_init = null
    
    [initialize.components]
    
    [initialize.tokenizer]
then I replaced the dafault NER pipeline with the trained one:

    nlp = it_core_news_lg.load()     
    nlp.remove_pipe("ner")    
    nlp.add_pipe("ner", source=spacy.load('training/model-best'))    

but when I try to perform DEP and POS tasks I got this error:
    
    [WARNING] [W113] Sourced component 'ner' may not work as expected: source vectors are not identical to current pipeline vectors.

**Comment by polm** [comment, marked as answer]
When you trained your NER model, it updated the tok2vec model and the NER model. So your NER model relies on the input of the tok2vec, which is not the same as the stock tok2vec. The warning is telling you that. Because the tok2vecs are different it is very likely your NER model won't work well with it. 

Your options are:

- train all the models you want to use with one shared tok2vec
- train the NER with a non-shared tok2vec, and include both in the pipeline

The latter options is probably easier and the right one for this situation, see the [section on non-shared tok2vecs](https://spacy.io/usage/training#:~:text=shared%20tok2vec%20listener%20layer) in the docs for how to set that up.

**Reply by adicrescenzo** [reply]
I tried the approach you suggested, this is my new config file:

    [paths]
    train = null
    dev = null
    vectors = null
    init_tok2vec = null
    
    [system]
    gpu_allocator = null
    seed = 0
    
    [nlp]
    lang = "it"
    pipeline = ["tok2vec","tagger","parser","ner"]
    batch_size = 1000
    disabled = []
    before_creation = null
    after_creation = null
    after_pipeline_creation = null
    tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
    
    [components]
    
    [components.tagger]
    source = "it_core_news_lg"
    replace_listeners = ["model.tok2vec"]
    
    [components.parser]
    source = "it_core_news_lg"
    replace_listeners = ["model.tok2vec"]
    
    [components.ner]
    factory = "ner"
    moves = null
    update_with_oracle_cut_size = 100
    
    [components.ner.model]
    @architectures = "spacy.TransitionBasedParser.v2"
    state_type = "ner"
    extra_state_tokens = false
    hidden_width = 64
    maxout_pieces = 2
    use_upper = true
    nO = null
    
    [components.ner.model.tok2vec]
    @architectures = "spacy.Tok2VecListener.v1"
    width = ${components.tok2vec.model.encode.width}
    upstream = "*"
    
    [components.tok2vec]
    source = "it_core_news_lg"
    replace_listeners = ["model.tok2vec"]
    #factory = "tok2vec"
    
    [components.tok2vec.model]
    @architectures = "spacy.Tok2Vec.v2"
    
    [components.tok2vec.model.embed]
    @architectures = "spacy.MultiHashEmbed.v2"
    width = ${components.tok2vec.model.encode.width}
    attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
    rows = [5000,2500,2500,2500]
    include_static_vectors = true
    
    [components.tok2vec.model.encode]
    @architectures = "spacy.MaxoutWindowEncoder.v2"
    width = 256
    depth = 8
    window_size = 1
    maxout_pieces = 3
    
    [corpora]
    
    [corpora.dev]
    @readers = "spacy.Corpus.v1"
    path = ${paths.dev}
    max_length = 0
    gold_preproc = false
    limit = 0
    augmenter = null
    
    [corpora.train]
    @readers = "spacy.Corpus.v1"
    path = ${paths.train}
    max_length = 2000
    gold_preproc = false
    limit = 0
    augmenter = null
    
    [training]
    dev_corpus = "corpora.dev"
    train_corpus = "corpora.train"
    seed = ${system.seed}
    gpu_allocator = ${system.gpu_allocator}
    dropout = 0.1
    accumulate_gradient = 1
    patience = 1600
    max_epochs = 0
    max_steps = 20000
    eval_frequency = 200
    frozen_components = ["tagger","parser"]
    before_to_disk = null
    
    [training.batcher]
    @batchers = "spacy.batch_by_words.v1"
    discard_oversize = false
    tolerance = 0.2
    get_length = null
    
    [training.batcher.size]
    @schedules = "compounding.v1"
    start = 100
    stop = 1000
    compound = 1.001
    t = 0.0
    
    [training.logger]
    @loggers = "spacy.ConsoleLogger.v1"
    progress_bar = false
    
    [training.optimizer]
    @optimizers = "Adam.v1"
    beta1 = 0.9
    beta2 = 0.999
    L2_is_weight_decay = true
    L2 = 0.01
    grad_clip = 1.0
    use_averages = false
    eps = 0.00000001
    learn_rate = 0.001
    
    [training.score_weights]
    ents_per_type = null
    ents_f = 1.0
    ents_p = 0.0
    ents_r = 0.0
    
    [pretraining]
    
    [initialize]
    vectors = "it_core_news_lg"
    init_tok2vec = ${paths.init_tok2vec}
    vocab_data = null
    lookups = null
    before_init = null
    after_init = null
    
    [initialize.components]
    
    [initialize.tokenizer]

But when I try to train the pipeline I get this error:

    ValueError: operands could not be broadcast together with shapes (7,96) (7,256) (7,96)

Am I missing something else? Thanks

**Reply by polm** [reply]
You set the width of your tok2vec component to 256, but other pretrained parts of the pipeline only have a width of 96.

```
[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 256 # change this to 96 and it will train
```

Did you change that for some reason? I think the quickstart will give a value of 96 there...

**Reply by adicrescenzo** [reply]
I left the default value created by the command `spacy init fill-config`.
Changing that value to 96 the training starts but seems that tok2vec doesn't converge.

![image](https://user-images.githubusercontent.com/6871290/119795912-3289bb80-bed9-11eb-8ebb-4fab65524edd.png)

furthermore if I use the trained pipeline I get the same error:

    [WARNING] [W113] Sourced component 'ner' may not work as expected: source vectors are not identical to current pipeline vectors.

Do you have any hint?

**Reply by polm** [reply]
OK, I think I figured out the issue. The width in tok2vec ended up being a distraction, you can change it back. 

The issue is this section:

```
[components.tok2vec]
source = "it_core_news_lg"
replace_listeners = ["model.tok2vec"]
#factory = "tok2vec"
```

It should look like this:

```
[components.tok2vec]
factory = "tok2vec"
```

To use the large model as a base, set the `init_tok2vec` value to `it_core_news_lg`. I was able to train and evaluate a model with that. (I didn't check the numbers, I just did the minimal training to verify everything worked.)

To explain what's happening, with the old settings your tok2vec component will be loaded wholesale from the pretrained pipeline every time you load your pipeline. So while it's updated during training, the **non-trained version** would be loaded due to your `source` argument. I'm also not sure what having `replace_listeners` on the tok2vec component will do, but since it doesn't listen to itself I don't think it does anything.
